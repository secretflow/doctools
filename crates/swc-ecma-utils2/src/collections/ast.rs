use swc_core::{
  common::util::take::Take,
  ecma::ast::{
    ArrayLit, CallExpr, Callee, ComputedPropName, Expr, ExprOrSpread, Ident, KeyValueProp, Lit,
    ObjectLit, Prop, PropName, PropOrSpread, Str,
  },
};

use super::{
  abc::{
    DefaultContainer, MappingBase, MutableMappingBase, MutableSequence, MutableSequenceBase,
    SequenceBase,
  },
  Sequence,
};

impl MappingBase for ObjectLit {
  type Key = Lit;
  type Value = Expr;

  fn _len(&self) -> usize {
    self._iter().count()
  }

  fn _iter<'k>(&'k self) -> Box<dyn Iterator<Item = Self::Key> + 'k> {
    Box::new(self.props.iter().filter_map(prop_name_to_lit))
  }

  fn _get(&self, key: &Self::Key) -> Option<&Self::Value> {
    self.props.iter().find_map(|prop| {
      if test_object_key(prop, key) {
        Some(&*prop.as_prop()?.as_key_value()?.value)
      } else {
        None
      }
    })
  }
}

impl MutableMappingBase for ObjectLit {
  fn _get_mut(&mut self, key: &Self::Key) -> Option<&mut Self::Value> {
    self.props.iter_mut().find_map(|prop| {
      if test_object_key(prop, key) {
        Some(&mut *prop.as_mut_prop()?.as_mut_key_value()?.value)
      } else {
        None
      }
    })
  }

  fn _set(&mut self, key: &Self::Key, value: Self::Value) {
    if let Some(found) = self._get_mut(key) {
      *found = value;
    } else {
      self.props.push(PropOrSpread::Prop(
        Prop::KeyValue(KeyValueProp {
          key: lit_to_prop_name(key),
          value: Box::new(value),
        })
        .into(),
      ));
    }
  }

  fn _del(&mut self, key: &Self::Key) -> Option<Self::Value> {
    let found = self.props.iter_mut().enumerate().find_map(|(idx, prop)| {
      if test_object_key(prop, key) {
        Some((
          idx,
          (&mut *prop.as_mut_prop()?.as_mut_key_value()?.value).take(),
        ))
      } else {
        None
      }
    });

    if let Some((idx, found)) = found {
      self.props.swap_remove(idx);
      Some(found)
    } else {
      None
    }
  }

  fn _pop(&mut self) -> Option<(Self::Key, Self::Value)> {
    let key = self.props.iter().rev().find_map(prop_name_to_lit)?;
    let value = self._del(&key)?;
    Some((key, value))
  }
}

fn prop_name_to_lit(prop: &PropOrSpread) -> Option<Lit> {
  match prop {
    PropOrSpread::Prop(prop) => match &**prop {
      Prop::KeyValue(KeyValueProp { key, .. }) => match &*key {
        PropName::Str(string) => Some(string.clone().into()),
        PropName::Num(number) => Some(number.clone().into()),
        PropName::Ident(ident) => Some(Str::from(&*ident.sym).into()),
        PropName::Computed(ComputedPropName { expr, .. }) => match &**expr {
          Expr::Lit(lit) => Some(lit.clone()),
          Expr::Ident(ident) => Some(Str::from(&*ident.sym).into()),
          _ => None,
        },
        _ => None,
      },
      _ => None,
    },
    _ => None,
  }
}

fn lit_to_prop_name(lit: &Lit) -> PropName {
  match lit {
    Lit::Str(string) => PropName::Str(string.clone()),
    Lit::Num(number) => PropName::Num(number.clone()),
    _ => unreachable!(),
  }
}

fn get_object_key_for_cmp(key: &Lit) -> Option<String> {
  match key {
    Lit::Str(string) => Some(string.value.to_string()),
    Lit::Num(number) => Some(number.value.to_string()),
    _ => None,
  }
}

fn test_object_key(prop: &PropOrSpread, test: &Lit) -> bool {
  let Some(key) = prop_name_to_lit(prop) else {
    return false;
  };
  let Some(key) = get_object_key_for_cmp(&key) else {
    return false;
  };
  let Some(test) = get_object_key_for_cmp(test) else {
    return false;
  };
  key == test
}

impl DefaultContainer for ObjectLit {
  fn create() -> Option<Self> {
    Some(Self::guarantee())
  }

  fn guarantee() -> Self {
    Self {
      span: Default::default(),
      props: vec![],
    }
  }
}

impl SequenceBase for ArrayLit {
  type Value = Expr;

  fn _len(&self) -> usize {
    self.elems.len()
  }

  fn _get(&self, idx: usize) -> Option<&Self::Value> {
    Some(self.elems.get(idx)?.as_ref()?.expr.as_ref())
  }
}

impl MutableSequenceBase for ArrayLit {
  fn _get_mut(&mut self, idx: usize) -> Option<&mut Self::Value> {
    Some(self.elems.get_mut(idx)?.as_mut()?.expr.as_mut())
  }

  fn _set(&mut self, idx: usize, value: Self::Value) -> Result<(), ()> {
    if idx >= self.elems.len() {
      Err(())
    } else {
      self.elems[idx] = Some(ExprOrSpread {
        spread: None,
        expr: Box::new(value),
      });
      Ok(())
    }
  }

  fn _del(&mut self, idx: usize) -> Option<Self::Value> {
    if idx >= self.elems.len() {
      None
    } else {
      let found = self.elems.remove(idx)?;
      Some(*found.expr)
    }
  }

  fn _ins(&mut self, idx: usize, value: Self::Value) {
    let idx = idx.min(self.elems.len());
    self.elems.insert(
      idx,
      Some(ExprOrSpread {
        spread: None,
        expr: Box::new(value),
      }),
    );
  }
}

impl DefaultContainer for ArrayLit {
  fn create() -> Option<Self> {
    Some(Self::guarantee())
  }

  fn guarantee() -> Self {
    Self {
      span: Default::default(),
      elems: vec![],
    }
  }
}

impl MappingBase for CallExpr {
  type Key = usize;
  type Value = Expr;

  fn _len(&self) -> usize {
    self.args.len() + 1
  }

  fn _iter<'k>(&'k self) -> Box<dyn Iterator<Item = Self::Key> + 'k> {
    Box::new(0..self._len())
  }

  fn _get(&self, key: &Self::Key) -> Option<&Self::Value> {
    if *key == 0usize {
      match &self.callee {
        Callee::Expr(ref expr) => Some(expr.as_ref()),
        _ => None,
      }
    } else {
      Some(self.args.get(key - 1)?.expr.as_ref())
    }
  }
}

impl MutableMappingBase for CallExpr {
  fn _get_mut(&mut self, key: &Self::Key) -> Option<&mut Self::Value> {
    if *key == 0usize {
      match &mut self.callee {
        Callee::Expr(ref mut expr) => Some(expr.as_mut()),
        _ => None,
      }
    } else {
      Some(self.args.get_mut(key - 1)?.expr.as_mut())
    }
  }

  fn _set(&mut self, key: &Self::Key, value: Self::Value) {
    if *key == 0 {
      self.callee = Callee::Expr(Box::new(value));
    } else {
      let idx = key - 1;
      if idx >= self.args.len() {
        self.args.resize_with(idx + 1, || ExprOrSpread {
          spread: None,
          expr: undefined().into(),
        });
      }
      self.args[idx] = ExprOrSpread {
        spread: None,
        expr: Box::new(value),
      };
    }
  }

  fn _del(&mut self, key: &Self::Key) -> Option<Self::Value> {
    if *key == 0 {
      let found = self.callee.take();
      let expr = *found.expr()?;
      Some(expr)
    } else if *key - 1 >= self.args.len() {
      None
    } else {
      let found = self.args.remove(*key - 1);
      let expr = *found.expr;
      Some(expr)
    }
  }

  fn _pop(&mut self) -> Option<(Self::Key, Self::Value)> {
    let idx = self._len().saturating_sub(1);
    Some((idx, self._del(&idx)?))
  }
}

fn undefined() -> Expr {
  Ident::new("undefined".into(), Default::default()).into()
}

impl DefaultContainer for CallExpr {
  fn create() -> Option<Self> {
    Some(Self::guarantee())
  }

  fn guarantee() -> Self {
    Self {
      span: Default::default(),
      callee: Callee::Expr(Box::new(Expr::dummy())),
      args: vec![],
      type_args: None,
    }
  }
}

impl MappingBase for Expr {
  type Key = Lit;
  type Value = Expr;

  fn _len(&self) -> usize {
    match self {
      Expr::Object(ref obj) => obj._len(),
      Expr::Array(ref arr) => arr.len(),
      Expr::Call(ref call) => call._len(),
      _ => 0,
    }
  }

  fn _iter<'k>(&'k self) -> Box<dyn Iterator<Item = Self::Key> + 'k> {
    match self {
      Expr::Object(ref obj) => obj._iter(),
      Expr::Array(ref arr) => Box::new((0..arr.len()).map(|idx| idx.to_string().into())),
      Expr::Call(ref call) => Box::new((0..call._len()).map(|idx| idx.to_string().into())),
      _ => Box::new(std::iter::empty()),
    }
  }

  fn _get(&self, key: &Self::Key) -> Option<&Self::Value> {
    match self {
      Expr::Object(ref obj) => obj._get(key),
      Expr::Array(ref arr) => lit_as_usize(key).and_then(|idx| arr.get_item(idx)),
      Expr::Call(ref call) => lit_as_usize(key).and_then(|idx| call._get(&idx)),
      _ => None,
    }
  }
}

impl MutableMappingBase for Expr {
  fn _get_mut(&mut self, key: &Self::Key) -> Option<&mut Self::Value> {
    match self {
      Expr::Object(ref mut obj) => obj._get_mut(key),
      Expr::Array(ref mut arr) => lit_as_usize(key).and_then(|idx| arr._get_mut(idx)),
      Expr::Call(ref mut call) => lit_as_usize(key).and_then(|idx| call._get_mut(&idx)),
      _ => None,
    }
  }

  fn _set(&mut self, key: &Self::Key, value: Self::Value) {
    match self {
      Expr::Object(ref mut obj) => obj._set(key, value),
      Expr::Array(ref mut arr) => {
        let idx = lit_as_usize(key).unwrap_or(arr.len());
        arr._set(idx, value).unwrap();
      }
      Expr::Call(ref mut call) => {
        let idx = lit_as_usize(key).unwrap_or(call._len());
        call._set(&idx, value);
      }
      _ => {}
    }
  }

  fn _del(&mut self, key: &Self::Key) -> Option<Self::Value> {
    match self {
      Expr::Object(ref mut obj) => obj._del(key),
      Expr::Array(ref mut arr) => {
        let idx = lit_as_usize(key).unwrap_or(arr.len());
        Some(arr._del(idx)?)
      }
      Expr::Call(ref mut call) => {
        let idx = lit_as_usize(key).unwrap_or(call._len());
        Some(call._del(&idx)?)
      }
      _ => None,
    }
  }

  fn _pop(&mut self) -> Option<(Self::Key, Self::Value)> {
    match self {
      Expr::Object(ref mut obj) => obj._pop(),
      Expr::Array(ref mut arr) => {
        let idx = arr.len().saturating_sub(1);
        Some((idx.to_string().into(), arr.pop(None)?))
      }
      Expr::Call(ref mut call) => call
        ._pop()
        .and_then(|(idx, expr)| Some((idx.to_string().into(), expr))),
      _ => None,
    }
  }
}

fn lit_as_usize(lit: &Lit) -> Option<usize> {
  match lit {
    Lit::Num(num) => {
      if num.value.fract() != 0.0 {
        None
      } else if num.value.is_sign_negative() {
        None
      } else if num.value > usize::MAX as f64 {
        None
      } else {
        Some(num.value as usize)
      }
    }
    Lit::Str(str) => str.value.parse().ok(),
    _ => None,
  }
}

impl DefaultContainer for Expr {
  fn create() -> Option<Self> {
    Some(Self::guarantee())
  }

  fn guarantee() -> Self {
    Expr::Object(ObjectLit::guarantee())
  }
}
