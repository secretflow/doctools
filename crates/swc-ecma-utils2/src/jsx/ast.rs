use swc_core::{
  common::util::take::Take as _,
  ecma::ast::{CallExpr, Expr, Ident, Lit, ObjectLit, Str},
};

use crate::ecma::itertools::is_nullish;

use super::{
  element::{JSXCall, JSXCallMut},
  tag::JSXTag,
  JSXRuntime,
};

impl<R: JSXRuntime> JSXCall<R> for CallExpr {
  const KEY_FACTORY: Self::Key = 0;
  const KEY_TYPE: Self::Key = 1;
  const KEY_PROPS: Self::Key = 2;

  type Component = Expr;
  type Props = ObjectLit;

  fn as_factory(value: &Self::Value) -> Option<&str> {
    let ident = value.as_ident()?;
    Some(&*ident.sym)
  }

  fn as_type(value: &Self::Value) -> Option<&Self::Component> {
    match value {
      Expr::Ident(_) => Some(value),
      Expr::Lit(Lit::Str(_)) => Some(value),
      _ => None,
    }
  }

  fn as_props(value: &Self::Value) -> Option<&Self::Props> {
    value.as_object()
  }

  fn as_tag(value: &Self::Value) -> Option<JSXTag> {
    match value {
      Expr::Lit(Lit::Str(Str { value, .. })) => Some(JSXTag::Intrinsic((&**value).into())),
      Expr::Ident(Ident { sym, .. }) => {
        if is_nullish(value) {
          None
        } else if sym == R::FRAGMENT {
          Some(JSXTag::Fragment)
        } else {
          Some(JSXTag::Component((&**sym).into()))
        }
      }
      _ => None,
    }
  }

  fn new() -> Self {
    CallExpr::dummy()
  }

  fn new_factory(factory: &str) -> Self::Value {
    Ident::from(factory).into()
  }

  fn new_fragment(fragment: &str) -> Self::Value {
    Ident::from(fragment).into()
  }

  fn new_props() -> Self::Value {
    ObjectLit::dummy().into()
  }
}

impl<R: JSXRuntime> JSXCallMut<R> for CallExpr {
  fn as_type_mut(value: &mut Self::Value) -> Option<&mut Self::Component> {
    match value {
      Expr::Ident(_) => Some(value),
      Expr::Lit(Lit::Str(_)) => Some(value),
      _ => None,
    }
  }

  fn as_props_mut(value: &mut Self::Value) -> Option<&mut Self::Props> {
    value.as_mut_object()
  }
}
