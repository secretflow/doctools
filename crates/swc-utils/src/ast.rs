use swc_core::{
  common::util::take::Take,
  ecma::{
    ast::{
      ArrayLit, CallExpr, Expr, ExprOrSpread, Ident, KeyValueProp, Lit, Number, ObjectLit, Prop,
      PropName, PropOrSpread, Str,
    },
    visit::{VisitMut, VisitMutWith},
  },
};

use crate::json::null;

#[derive(Debug)]
pub enum PropVisitorError {
  InvalidPath(String),
  NotFound,
}

struct PropVisitor<'path, T> {
  path: &'path [Lit],
  ensure: bool,
  idx: usize,
  callback: &'path mut dyn FnMut(&mut Box<Expr>) -> T,
  result: Result<T, PropVisitorError>,
}

macro_rules! item_getter_error {
    ($this:ident, $error:path, $msg:literal $(, $arg:expr)*) => {{
        $this.result = Err($error(format!($msg, $($arg),*).to_string()));
        return;
    }};
    ($this:ident, $error:path) => {{
        $this.result = Err($error);
        return;
    }};
}

impl<T> PropVisitor<'_, T> {
  fn traverse<N: VisitMutWith<Self>>(&mut self, expr: &mut N) {
    self.idx += 1;
    expr.visit_mut_with(self);
  }

  fn found(&mut self, found: &mut Box<Expr>) {
    self.result = Ok((self.callback)(found));
  }

  fn process_array_elems(&mut self, elems: &mut Vec<Option<ExprOrSpread>>) {
    let subscript = self.path.get(self.idx).expect("index out of bounds");

    let index = match subscript {
      Lit::Num(num) => match f64_to_usize(num.value) {
        Some(idx) => idx,
        None => item_getter_error!(
          self,
          PropVisitorError::InvalidPath,
          "invalid array index {}",
          num.value
        ),
      },
      Lit::Str(str) => match str.value.parse::<usize>() {
        Ok(idx) => idx,
        Err(_) => item_getter_error!(
          self,
          PropVisitorError::InvalidPath,
          "invalid array index {}",
          str.value
        ),
      },
      _ => item_getter_error!(
        self,
        PropVisitorError::InvalidPath,
        "array index should be Lit::Num, found {:?}",
        subscript
      ),
    };

    let mut set_default = |arr: &mut Vec<Option<ExprOrSpread>>| {
      if self.ensure {
        if index >= arr.len() {
          arr.resize_with(index + 1, || None);
        }
        arr[index] = Some(ExprOrSpread {
          spread: None,
          expr: {
            let mut new = Box::from(null());
            (self.callback)(&mut new);
            new
          },
        })
      } else {
        item_getter_error!(self, PropVisitorError::NotFound)
      }
    };

    let ensure_object = || {
      Some(ExprOrSpread {
        spread: None,
        expr: Box::from(ObjectLit {
          props: vec![],
          span: Default::default(),
        }),
      })
    };

    if self.idx == self.path.len() - 1 {
      match elems.get_mut(index) {
        Some(item) => match item {
          Some(ExprOrSpread { expr, .. }) => {
            self.found(expr);
          }
          None => set_default(elems),
        },
        None => set_default(elems),
      }
    } else {
      match elems.get_mut(index) {
        Some(next) => match next {
          Some(ExprOrSpread { expr, .. }) => self.traverse(expr),
          None => {
            if self.ensure {
              *next = ensure_object();
              self.traverse(next.as_mut().unwrap().expr.as_mut());
            } else {
              item_getter_error!(self, PropVisitorError::NotFound)
            }
          }
        },
        None => {
          if self.ensure {
            elems.resize_with(index + 1, || None);
            elems[index] = ensure_object();
            self.traverse(elems[index].as_mut().unwrap().expr.as_mut());
          } else {
            item_getter_error!(self, PropVisitorError::NotFound)
          }
        }
      };
    }
  }

  fn process_object_props(&mut self, props: &'_ mut Vec<PropOrSpread>) {
    let subscript = self.path.get(self.idx).expect("index out of bounds");

    let key = match subscript {
      Lit::Str(key) => key.value.to_string(),
      Lit::Num(num) => num.value.to_string(),
      _ => item_getter_error!(
        self,
        PropVisitorError::InvalidPath,
        "object keys should be strings or numbers, found {:?}",
        subscript
      ),
    };

    let key_eq = prop_is_named(&key);

    let existing = props.iter_mut().find_map(|p| {
      key_eq(p)
        .then(|| p.as_mut_prop().unwrap())
        .and_then(|p| p.as_mut_key_value())
    });

    if self.idx == self.path.len() - 1 {
      match existing {
        Some(KeyValueProp { value, .. }) => {
          self.found(value);
        }
        None => {
          if self.ensure {
            props.push(
              PropOrSpread::Prop(
                Prop::from(KeyValueProp {
                  key: PropName::Str(key.as_str().into()),
                  value: {
                    let mut new = Box::from(null());
                    (self.callback)(&mut new);
                    new
                  },
                })
                .into(),
              )
              .into(),
            )
          } else {
            item_getter_error!(self, PropVisitorError::NotFound)
          }
        }
      }
    } else {
      match existing {
        Some(prop) => self.traverse(&mut prop.value),
        None => {
          if self.ensure {
            props.push(
              PropOrSpread::Prop(
                Prop::from(KeyValueProp {
                  key: PropName::Str(key.as_str().into()),
                  value: Box::from(ObjectLit {
                    props: vec![],
                    span: Default::default(),
                  }),
                })
                .into(),
              )
              .into(),
            );
            self.traverse(
              props
                .last_mut()
                .unwrap()
                .as_mut_prop()
                .unwrap()
                .as_mut_key_value()
                .unwrap()
                .value
                .as_mut(),
            );
          } else {
            item_getter_error!(self, PropVisitorError::NotFound)
          }
        }
      }
    }
  }
}

impl<T> VisitMut for PropVisitor<'_, T> {
  fn visit_mut_array_lit(&mut self, array: &mut ArrayLit) {
    self.process_array_elems(&mut array.elems);
  }

  fn visit_mut_object_lit(&mut self, object: &mut ObjectLit) {
    self.process_object_props(&mut object.props)
  }

  fn visit_mut_call_expr(&mut self, call: &mut CallExpr) {
    let mut args = call.args.drain(..).map(Some).collect();
    self.process_array_elems(&mut args);
    call.args = args.drain(..).map(Option::unwrap).collect();
  }
}

pub fn mut_call_by_path<T>(
  call: &mut CallExpr,
  path: &[Lit],
  mut callback: impl FnMut(&mut Box<Expr>) -> T,
) -> Result<T, PropVisitorError> {
  let mut setter = PropVisitor {
    path,
    idx: 0,
    ensure: false,
    // This cannot be boxed because of the lifetime.
    //
    // With Box<dyn FnMut>, rustc asks for callback to be 'static
    // which limits what kind of closures can be passed
    // (essentially cannot borrow anything whose lifetime is shorter than 'static).
    //
    // I still don't know why Box requires 'static though.
    callback: &mut callback,
    result: Err(PropVisitorError::NotFound),
  };
  call.visit_mut_with(&mut setter);
  setter.result
}

pub fn mut_ast_by_path<T>(
  ast: &mut Expr,
  path: &[Lit],
  mut callback: impl FnMut(&mut Box<Expr>) -> T,
) -> Result<T, PropVisitorError> {
  let mut setter = PropVisitor {
    path,
    idx: 0,
    ensure: false,
    callback: &mut callback,
    result: Err(PropVisitorError::NotFound),
  };
  if let Some(object) = ast.as_mut_object() {
    object.visit_mut_with(&mut setter)
  } else if let Some(array) = ast.as_mut_array() {
    array.visit_mut_with(&mut setter)
  } else if let Some(call) = ast.as_mut_call() {
    call.visit_mut_with(&mut setter)
  } else {
    unreachable!("unsupported expression {:?}", ast)
  };
  setter.result
}

pub fn set_ast_by_path(
  ast: &mut Expr,
  path: &[Lit],
  mut value: Box<Expr>,
) -> Result<(), PropVisitorError> {
  let mut setter = PropVisitor {
    path,
    idx: 0,
    ensure: true,
    callback: &mut move |expr: &mut Box<Expr>| *expr = value.take().into(),
    result: Ok(()),
  };
  if let Some(object) = ast.as_mut_object() {
    object.visit_mut_with(&mut setter)
  } else if let Some(array) = ast.as_mut_array() {
    array.visit_mut_with(&mut setter)
  } else if let Some(call) = ast.as_mut_call() {
    call.visit_mut_with(&mut setter)
  } else {
    unreachable!("unsupported expression {:?}", ast)
  };
  setter.result
}

fn prop_is_named(key: &str) -> impl Fn(&PropOrSpread) -> bool + '_ {
  move |prop: &PropOrSpread| match prop.as_prop().and_then(|p| p.as_key_value()) {
    None => false,
    Some(KeyValueProp { key: prop, .. }) => match prop {
      PropName::Str(Str { value, .. }) => &*value == key,
      PropName::Ident(Ident { sym, .. }) => &*sym == key,
      PropName::Num(Number { value, .. }) => value.to_string() == key,
      _ => false,
    },
  }
}

fn f64_to_usize(f: f64) -> Option<usize> {
  if f.is_finite() && f.fract() == 0.0 && f >= 0.0 && f <= usize::MAX as f64 {
    Some(f as usize)
  } else {
    None
  }
}

#[cfg(test)]
mod tests {
  use serde_json::json;
  use swc_core::{
    ecma::{
      ast::{CallExpr, Lit},
      codegen::Config,
    },
    testing::DebugUsingDisplay,
  };

  use super::{mut_ast_by_path, set_ast_by_path};
  use crate::{json::json_expr, testing::print_one};

  #[test]
  fn test_set_object() {
    let mut expr = json_expr(json!({}));

    set_ast_by_path(
      &mut expr,
      &[Lit::from("children")],
      Box::from(json_expr(json!([]))),
    )
    .unwrap();

    let code = print_one(&expr, None, Some(Config::default().with_minify(true)));
    assert_eq!(
      DebugUsingDisplay(code.unwrap().as_str()),
      DebugUsingDisplay(r#"{"children":[]}"#)
    );
  }

  #[test]
  fn test_set_object_deeply() {
    let mut expr = json_expr(json!({
        "lorem": {
            "ipsum": {
                "dolor": [
                    "sit",
                    "amet"
                ]
            }
        }
    }));

    set_ast_by_path(
      &mut expr,
      &[
        Lit::from("lorem"),
        Lit::from("ipsum"),
        Lit::from("dolor"),
        Lit::from(2),
      ],
      Box::from(json_expr(json!("consectetur adipiscing elit"))),
    )
    .unwrap();

    let code = print_one(&expr, None, Some(Config::default().with_minify(true)));

    assert_eq!(
      DebugUsingDisplay(code.unwrap().as_str()),
      DebugUsingDisplay(
        r#"{"lorem":{"ipsum":{"dolor":["sit","amet","consectetur adipiscing elit"]}}}"#
      )
    );
  }

  #[test]
  fn test_mut_function() {
    let call = serde_json::from_str::<CallExpr>(
            r#"
            {
                "type": "CallExpression",
                "span": {
                  "start": 0,
                  "end": 130,
                  "ctxt": 0
                },
                "callee": {
                  "type": "Identifier",
                  "span": {
                    "start": 0,
                    "end": 3,
                    "ctxt": 1
                  },
                  "value": "jsx",
                  "optional": false
                },
                "arguments": [
                  {
                    "spread": null,
                    "expression": {
                      "type": "StringLiteral",
                      "span": {
                        "start": 4,
                        "end": 7,
                        "ctxt": 0
                      },
                      "value": "a",
                      "raw": "\"a\""
                    }
                  },
                  {
                    "spread": null,
                    "expression": {
                      "type": "ObjectExpression",
                      "span": {
                        "start": 9,
                        "end": 129,
                        "ctxt": 0
                      },
                      "properties": [
                        {
                          "type": "KeyValueProperty",
                          "key": {
                            "type": "StringLiteral",
                            "span": {
                              "start": 10,
                              "end": 17,
                              "ctxt": 0
                            },
                            "value": "props",
                            "raw": "\"props\""
                          },
                          "value": {
                            "type": "ObjectExpression",
                            "span": {
                              "start": 19,
                              "end": 128,
                              "ctxt": 0
                            },
                            "properties": [
                              {
                                "type": "KeyValueProperty",
                                "key": {
                                  "type": "StringLiteral",
                                  "span": {
                                    "start": 20,
                                    "end": 26,
                                    "ctxt": 0
                                  },
                                  "value": "href",
                                  "raw": "\"href\""
                                },
                                "value": {
                                  "type": "CallExpression",
                                  "span": {
                                    "start": 28,
                                    "end": 127,
                                    "ctxt": 0
                                  },
                                  "callee": {
                                    "type": "Identifier",
                                    "span": {
                                      "start": 28,
                                      "end": 32,
                                      "ctxt": 1
                                    },
                                    "value": "_url",
                                    "optional": false
                                  },
                                  "arguments": [
                                    {
                                      "spread": null,
                                      "expression": {
                                        "type": "StringLiteral",
                                        "span": {
                                          "start": 33,
                                          "end": 43,
                                          "ctxt": 0
                                        },
                                        "value": "external",
                                        "raw": "\"external\""
                                      }
                                    },
                                    {
                                      "spread": null,
                                      "expression": {
                                        "type": "NullLiteral",
                                        "span": {
                                          "start": 45,
                                          "end": 49,
                                          "ctxt": 0
                                        }
                                      }
                                    },
                                    {
                                      "spread": null,
                                      "expression": {
                                        "type": "StringLiteral",
                                        "span": {
                                          "start": 51,
                                          "end": 126,
                                          "ctxt": 0
                                        },
                                        "value": "https://en.wikipedia.org/wiki/The_quick_brown_fox_jumps_over_the_lazy_dog",
                                        "raw": "\"https://en.wikipedia.org/wiki/The_quick_brown_fox_jumps_over_the_lazy_dog\""
                                      }
                                    }
                                  ],
                                  "typeArguments": null
                                }
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ],
                "typeArguments": null
              }
            "#,
        ).unwrap();

    let mut expr = Box::from(call);

    assert_eq!(
      mut_ast_by_path(
        &mut expr,
        &[
          Lit::from(1),
          Lit::from("props"),
          Lit::from("href"),
          Lit::from("2"),
        ],
        |expr| {
          *expr =
            Lit::from("https://fr.wikipedia.org/wiki/Portez_ce_vieux_whisky_au_juge_blond_qui_fume")
              .into()
        },
      )
      .unwrap(),
      ()
    );

    let code = print_one(&expr, None, Some(Config::default().with_minify(true)));

    assert_eq!(
      DebugUsingDisplay(code.unwrap().as_str()),
      DebugUsingDisplay(
        r#"jsx("a",{"props":{"href":_url("external",null,"https://fr.wikipedia.org/wiki/Portez_ce_vieux_whisky_au_juge_blond_qui_fume")}})"#
      )
    );
  }
}
