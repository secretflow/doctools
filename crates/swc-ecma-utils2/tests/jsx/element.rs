use swc_core::ecma::ast::Ident;

use swc_ecma_testing2::{insta, print_one_unchecked};
use swc_ecma_utils2::{
  collections::MutableMapping,
  jsx::{create_element, create_fragment, jsx_mut, JSXElementMut, JSXRuntimeDefault},
};

#[test]
fn test_fragment() {
  let document = print_one_unchecked(&create_fragment::<JSXRuntimeDefault>());
  insta::assert_snapshot!(document);
}

#[test]
fn test_intrinsic() {
  let document = print_one_unchecked(&{
    let mut elem = create_element::<JSXRuntimeDefault>("div".into());
    jsx_mut::<JSXRuntimeDefault>(&mut elem)
      .get_props_mut()
      .set_item("children", "foo".into());
    elem
  });
  insta::assert_snapshot!(document);
}

#[test]
fn test_component() {
  let document = print_one_unchecked(&create_element::<JSXRuntimeDefault>(
    Ident::new("Foo".into(), Default::default()).into(),
  ));
  insta::assert_snapshot!(document);
}

#[test]
fn test_props() {
  let document = print_one_unchecked(&{
    let mut elem = create_element::<JSXRuntimeDefault>("div".into());
    jsx_mut::<JSXRuntimeDefault>(&mut elem)
      .get_props_mut()
      .set_item("className", "foo".into())
      .set_item("id", "bar".into());
    elem
  });
  insta::assert_snapshot!(document);
}
