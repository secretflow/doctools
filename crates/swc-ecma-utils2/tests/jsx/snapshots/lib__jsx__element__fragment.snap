---
source: crates/swc-ecma-utils2/tests/jsx/element.rs
expression: document
---
_jsx(_Fragment, {})
